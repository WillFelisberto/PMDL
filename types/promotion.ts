export type ConditionType =
  | 'comparator'
  | 'iterator'
  | 'group-iterator'
  | 'quantifier'
  | 'operator'
  | 'operation';

export type ConditionBase = {
  id: string;
  type: ConditionType;
  attributes?: Array<{ name: string; value: string }>;
};

export type ComparatorCondition = ConditionBase & {
  type: 'comparator';
  comparatorType:
    | 'equals'
    | 'not-equals'
    | 'less-than'
    | 'less-than-or-equals'
    | 'greater-than'
    | 'greater-than-or-equals'
    | 'contains'
    | 'starts-with'
    | 'ends-with'
    | 'isoneof'
    | 'isnotoneof'
    | 'includes'
    | 'includes-any'
    | 'includes-all';
  leftValue: string;
  rightValue: string;
  dataType: string;
};

export type ValueType = {
  type: 'value' | 'constant' | 'array' | 'null';
  dataType?: string;
  value?: string | number | boolean;
  values?: ValueType[];
};
export type IteratorCondition = ConditionBase & {
  type: 'iterator';
  iteratorType: 'next' | 'every' | 'up-to-and-including' | 'custom';
  collectionName: string;
  elementName: string;
  number?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  condition?: PromotionCondition;
};

export type GroupIteratorCondition = ConditionBase & {
  type: 'group-iterator';
  iteratorType: 'every-group' | 'up-to-and-including-group' | 'custom';
  number?: number;
  conditions?: PromotionCondition[];
};

export type QuantifierCondition = ConditionBase & {
  type: 'quantifier';
  quantifierType: 'exactly' | 'at-least' | 'at-most' | 'all' | 'custom';
  number?: number;
  collectionName: string;
  elementName: string;
  condition?: PromotionCondition;
  customType?: string;
};

export type OperatorCondition = ConditionBase & {
  type: 'operator';
  operatorType: 'and' | 'or' | 'not' | 'custom';
  conditions: PromotionCondition[];
  customType?: string; // Para operadores personalizados
};

export type OperationCondition = ConditionBase & {
  type: 'operation';
  operationType: 'union' | 'anded-union' | 'custom';
  iterators: PromotionCondition[];
  customType?: string;
};
export type PromotionCondition =
  | ComparatorCondition
  | IteratorCondition
  | GroupIteratorCondition
  | QuantifierCondition
  | OperatorCondition
  | OperationCondition;

export type DiscountStructure = {
  calculatorType: string;
  discountType: 'percentOff' | 'amountOff' | 'fixedPrice';
  adjuster: number;
  target?: 'order' | 'shipping' | 'item';
};

export type Offer = {
  discountStructures: DiscountStructure[];
};

export type Promotion = {
  displayName: string;
  description: string;
  enabled: boolean;
  priority: number;
  qualifier: PromotionCondition | undefined;
  offer: Offer;
  conditions: PromotionCondition[];
  sites: Array<{ repositoryId: string }>;
};
