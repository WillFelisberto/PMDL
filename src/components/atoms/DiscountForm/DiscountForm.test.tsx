import { DiscountForm } from '.';

import { fireEvent, render, screen } from '@/tests/test-utils';
import { DiscountStructure } from '@/types/promotion';

describe('DiscountForm', () => {
  const mockOnChange = jest.fn();

  const defaultDiscount: DiscountStructure = {
    calculatorType: 'default',
    discountType: 'percentOff',
    adjuster: 10
  };

  beforeEach(() => {
    mockOnChange.mockClear();
  });

  it('should render correctly the form', () => {
    const { container } = render(
      <DiscountForm discount={defaultDiscount} onChange={mockOnChange} />
    );

    expect(screen.getByText('游눯 Configurar Desconto')).toBeInTheDocument();

    expect(screen.getByLabelText('Tipo de Desconto')).toBeInTheDocument();

    expect(screen.getByLabelText('Porcentagem')).toBeInTheDocument();

    expect(
      screen.getByText('O desconto ser치 aplicado como porcentagem sobre o total.')
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('Deve chamar onChange ao alterar o tipo de desconto', () => {
    render(<DiscountForm discount={defaultDiscount} onChange={mockOnChange} />);

    const select = screen.getByLabelText('Tipo de Desconto');

    fireEvent.change(select, { target: { value: 'amountOff' } });

    expect(mockOnChange).toHaveBeenCalledWith({
      ...defaultDiscount,
      discountType: 'amountOff'
    });

    fireEvent.change(select, { target: { value: 'fixedPrice' } });

    expect(mockOnChange).toHaveBeenCalledWith({
      ...defaultDiscount,
      discountType: 'fixedPrice'
    });
  });

  it('Deve chamar onChange ao alterar o valor do desconto', () => {
    render(<DiscountForm discount={defaultDiscount} onChange={mockOnChange} />);

    const input = screen.getByLabelText('Porcentagem');

    fireEvent.change(input, { target: { value: '25' } });

    expect(mockOnChange).toHaveBeenCalledWith({
      ...defaultDiscount,
      adjuster: 25
    });

    fireEvent.change(input, { target: { value: '30.5' } });

    expect(mockOnChange).toHaveBeenCalledWith({
      ...defaultDiscount,
      adjuster: 30.5
    });
  });

  it('Deve exibir corretamente os r칩tulos e dicas baseados no tipo de desconto', () => {
    const { rerender } = render(
      <DiscountForm discount={defaultDiscount} onChange={mockOnChange} />
    );

    expect(
      screen.getByText('O desconto ser치 aplicado como porcentagem sobre o total.')
    ).toBeInTheDocument();

    rerender(
      <DiscountForm
        discount={{ ...defaultDiscount, discountType: 'amountOff' }}
        onChange={mockOnChange}
      />
    );
    expect(screen.getByText('O valor fixo ser치 subtra칤do do total do pedido.')).toBeInTheDocument();

    rerender(
      <DiscountForm
        discount={{ ...defaultDiscount, discountType: 'fixedPrice' }}
        onChange={mockOnChange}
      />
    );
    expect(screen.getByText('O pre칞o final ser치 definido para este valor.')).toBeInTheDocument();
  });
});
