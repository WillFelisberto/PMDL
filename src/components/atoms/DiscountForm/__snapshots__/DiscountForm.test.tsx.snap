// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountForm should render correctly the form 1`] = `
<div>
  <div
    class="bg-gray-50 p-4 rounded-lg border"
  >
    <h3
      class="text-lg font-semibold text-gray-700 mb-4"
    >
      ðŸ’° Configurar Desconto
    </h3>
    <div
      class="grid grid-cols-1 md:grid-cols-2 gap-4"
    >
      <div
        class="space-y-2"
      >
        <label
          class="block text-sm font-medium text-gray-600"
          for="discountType"
        >
          Tipo de Desconto
        </label>
        <select
          class="w-full p-2 border rounded-md bg-white focus:ring-2 focus:ring-blue-500"
          id="discountType"
        >
          <option
            value="percentOff"
          >
            Porcentagem (%)
          </option>
          <option
            value="amountOff"
          >
            Valor Fixo (R$)
          </option>
          <option
            value="fixedPrice"
          >
            PreÃ§o Fixo
          </option>
        </select>
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="block text-sm font-medium text-gray-600"
          for="discountValue"
        >
          Porcentagem
        </label>
        <div
          class="relative"
        >
          <span
            class="absolute left-3 top-1/2 -translate-y-1/2 text-gray-500"
          >
            %
          </span>
          <input
            class="w-full p-2 pl-8 border rounded-md"
            id="discountValue"
            min="0"
            placeholder="Ex: 10"
            step="1"
            type="number"
            value="10"
          />
        </div>
      </div>
    </div>
    <p
      class="mt-3 text-sm text-gray-500"
    >
      O desconto serÃ¡ aplicado como porcentagem sobre o total.
    </p>
  </div>
</div>
`;
